tests/PHPStan/Parser/CachedParserTest.php

```diff
@@ -0,0 +1,123 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Parser;
+
+class CachedParserTest extends \PHPUnit\Framework\TestCase
+{
+
+	/**
+	 * @dataProvider dataParseFileClearCache
+	 * @param int $cachedNodesByFileCountMax
+	 * @param int $cachedNodesByStringCountMax
+	 * @param int $cachedNodesByFileCountExpected
+	 * @param int $cachedNodesByStringCountExpected
+	 */
+	public function testParseFileClearCache(
+		int $cachedNodesByFileCountMax,
+		int $cachedNodesByStringCountMax,
+		int $cachedNodesByFileCountExpected,
+		int $cachedNodesByStringCountExpected
+	): void
+	{
+		$parser = new CachedParser(
+			$this->getParserMock(),
+			$cachedNodesByFileCountMax,
+			$cachedNodesByStringCountMax
+		);
+
+		// Add files to cache
+		for ($i = 0; $i <= $cachedNodesByFileCountMax; $i++) {
+			$parser->parseFile('file' . $i);
+		}
+
+		$reflection = new \ReflectionClass($parser);
+		$reflectionCachedNodesByFileCount = $reflection->getProperty('cachedNodesByFileCount');
+		$reflectionCachedNodesByFileCount->setAccessible(true);
+
+		$this->assertEquals(
+			$cachedNodesByFileCountExpected,
+			$reflectionCachedNodesByFileCount->getValue($parser)
+		);
+
+		$reflectionCachedNodesByFile = $reflection->getProperty('cachedNodesByFile');
+		$reflectionCachedNodesByFile->setAccessible(true);
+
+		$this->assertCount(
+			$cachedNodesByFileCountExpected,
+			$reflectionCachedNodesByFile->getValue($parser)
+		);
+
+		// Add strings to cache
+		for ($i = 0; $i <= $cachedNodesByStringCountMax; $i++) {
+			$parser->parseString('string' . $i);
+		}
+
+		$reflection = new \ReflectionClass($parser);
+		$reflectionCachedNodesByStringCount = $reflection->getProperty('cachedNodesByStringCount');
+		$reflectionCachedNodesByStringCount->setAccessible(true);
+
+		$this->assertEquals(
+			$cachedNodesByStringCountExpected,
+			$reflectionCachedNodesByStringCount->getValue($parser)
+		);
+
+		$reflectionCachedNodesByString = $reflection->getProperty('cachedNodesByString');
+		$reflectionCachedNodesByString->setAccessible(true);
+
+		$this->assertCount(
+			$cachedNodesByStringCountExpected,
+			$reflectionCachedNodesByString->getValue($parser)
+		);
+	}
+
+	public function dataParseFileClearCache(): \Generator
+	{
+		yield 'even' => [
+			'cachedNodesByFileCountMax' => 100,
+			'cachedNodesByStringCountMax' => 50,
+			'cachedNodesByFileCountExpected' => 51,
+			'cachedNodesByStringCountExpected' => 26,
+		];
+
+		yield 'odd' => [
+			'cachedNodesByFileCountMax' => 101,
+			'cachedNodesByStringCountMax' => 51,
+			'cachedNodesByFileCountExpected' => 52,
+			'cachedNodesByStringCountExpected' => 27,
+		];
+	}
+
+	/**
+	 * @return Parser|\PHPUnit\Framework\MockObject\MockObject
+	 */
+	private function getParserMock()
```

Comment:
Better way is to do an intersection - Parser&MockObject.

Reply:
Done

Reply:
Forgot we have phpstan-phpunit